Implement softmax_backward

vtsilytskyiTT opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/27796

### Problem description
Lack of softmax_backward op.

### What's changed
Basic implementation of softmax_backward fused operation with the unit test.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

vtsilytskyiTT committed:
Initial softmax_backward implementation

Naive composite operation.

vtsilytskyiTT committed:
Some kernel impl, runtime error

philei-tt committed:
softmax backwards skeleton fix (#30166)

### Ticket
N/A

### What's changed
Fix fused softmax backward skeleton that computes eltwise multiply
- Wrong storage type in InterleavedAddrgen in writer (L1 instead of
DRAM). Replaced InterleavedAddrgen with TensorAccessor to avoid this
pitfall
- Wrongly used init for binary op and for mul_tiles
- Missing tile registers syncronization in compute
- Fix underflow in computation of num_tiles_per_core
- Fix negative dim case (underflow)
- Some other small fixes that could result in an error later

Still fails for non-trivial shapes @vtsilytskyiTT this is up to you to
debug
@vtsilytskyiTT please ping me if you have some questions

vtsilytskyiTT committed:
Add data types, revert -1 limitation

vtsilytskyiTT committed:
Some checkpoint, compiles but test fails

vtsilytskyiTT committed:
Print tensors into file for debug purpose

rfurko-tt commented on ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/softmax_backward_program_factory.cpp:
Querying runtime args can be quite inefficient, consider optimizing this code this way
```
auto reader_runtime_args = GetRuntimeArgs(program, reader_kernel_id);
for (....) {
   auto& reader_runtime_args_per_core = reader_runtime_args[core.x][core.y];
   .... 
}
```

vtsilytskyiTT replied:
Done.

rfurko-tt commented on ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp:
please rename cb handles in a proper way. `y_cb_id`, `grad_cb_id`, `grad_minus_sum_cb_id` etc
it would be so much easier to read the code

rfurko-tt commented on ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/dataflow/reader_softmax_backward.cpp:
why don't you use common method? 
```
noc_async_read_tile
```

vtsilytskyiTT commented on ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/softmax_backward_device_operation.cpp:
Done

vtsilytskyiTT commented on ttnn/cpp/ttnn/operations/normalization/softmax_backward/softmax_backward.cpp:
Indeed, it is redundant. Removed.

vtsilytskyiTT committed:
Address some of the comments

bbradelTT commented on ttnn/cpp/ttnn/operations/normalization/softmax_backward/device/kernels/compute/softmax_backward_kernel.cpp:
```diff
@@ -0,0 +1,174 @@
+    sub_tiles_init(src1_cb_id, sum_reduce_cb_id);
+
+    tile_regs_acquire();
+    sub_tiles(src1_cb_id, sum_reduce_cb_id, src1_tile_idx, 0, 0);  // grad[w] - sum[0]
```
I think this is an elementwise sub, which won't work because sum_reduce_cb_id has only the first column populated properly. You should look into using a broadcast sub.

vtsilytskyiTT committed:
Finally accurate result!

Use correct subtraction in compute kernel.
Fix -1 dimension handling.
Enable BFLOAT8, which just works for free.

vtsilytskyiTT committed:
Polishing

Increase tensor int the test to exceed L1 memory size.
Extract debug code into functions.
More arguments in ComputeConfig; doesn't improve accuracy though.

vtsilytskyiTT committed:
Rebase branch, move the test

vtsilytskyiTT committed:
Temporary changes for debug purpose

vtsilytskyiTT committed:
Use matmul instead of reduce

Accuracy is 10 times better.

vtsilytskyiTT commented:
Closing in favor of https://github.com/tenstorrent/tt-metal/pull/31580, production-ready implementation.
